### PR TITLE
support sparse opnorm

### PR DESCRIPTION
**Pull request by Roger-luo**
I only supported Inf opnorm, which is relatively simple, and is used in Krylov subspace methods. I guess it's fine to just error on other norms so if anyone hit it they could submit PR for it.

### CODE REVIEW

**Comment by Roger-luo** [comment]
A side question: maybe GPUArrays should have some kind of sparse abstraction so that this code can be generic?

**Comment by Roger-luo** [comment]
I realize somehow this kernel doesn't work on large sparse array, and I got the following

```julia
n = 15
A = sprand(1<<n, 1<<n, 0.01)
dA = cu(A)
```

Any idea why it errors in this way? 

```julia
julia> sum_dim1(CuSparseMatrixCSR(dA))
ERROR: CUDA error: invalid argument (code 1, ERROR_INVALID_VALUE)
Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA ~/.julia/packages/CUDA/5jdFl/lib/cudadrv/error.jl:91
  [2] macro expansion
    @ ~/.julia/packages/CUDA/5jdFl/lib/cudadrv/error.jl:101 [inlined]
  [3] cuLaunchKernel(f::CuFunction, gridDimX::UInt32, gridDimY::UInt32, gridDimZ::UInt32, blockDimX::UInt32, blockDimY::UInt32, blockDimZ::UInt32, sharedMemBytes::Int64, hStream::CuStream, kernelParams::Vector{Ptr{Nothing}}, extra::Ptr{Nothing})
    @ CUDA ~/.julia/packages/CUDA/5jdFl/lib/utils/call.jl:26
  [4] #39
    @ ~/.julia/packages/CUDA/5jdFl/lib/cudadrv/execution.jl:69 [inlined]
  [5] macro expansion
    @ ~/.julia/packages/CUDA/5jdFl/lib/cudadrv/execution.jl:33 [inlined]
  [6] macro expansion
    @ ./none:0 [inlined]
  [7] pack_arguments(::CUDA.var"#39#40"{Bool, Int64, CuStream, CuFunction, CuDim3, CuDim3}, ::CUDA.KernelState, ::CuDeviceVector{Float32, 1}, ::CuSparseDeviceMatrixCSR{Float32, Int32, 1})
    @ CUDA ./none:0
  [8] #launch#38
    @ ~/.julia/packages/CUDA/5jdFl/lib/cudadrv/execution.jl:62 [inlined]
  [9] #44
    @ ~/.julia/packages/CUDA/5jdFl/lib/cudadrv/execution.jl:136 [inlined]
 [10] macro expansion
    @ ~/.julia/packages/CUDA/5jdFl/lib/cudadrv/execution.jl:95 [inlined]
 [11] macro expansion
    @ ./none:0 [inlined]
 [12] convert_arguments
    @ ./none:0 [inlined]
 [13] #cudacall#43
    @ ~/.julia/packages/CUDA/5jdFl/lib/cudadrv/execution.jl:135 [inlined]
 [14] macro expansion
    @ ~/.julia/packages/CUDA/5jdFl/src/compiler/execution.jl:204 [inlined]
 [15] macro expansion
    @ ./none:0 [inlined]
 [16] #call#241
    @ ./none:0 [inlined]
 [17] #_#262
    @ ~/.julia/packages/CUDA/5jdFl/src/compiler/execution.jl:462 [inlined]
 [18] macro expansion
    @ ~/.julia/packages/CUDA/5jdFl/src/compiler/execution.jl:104 [inlined]
 [19] sum_dim1(A::CuSparseMatrixCSR{Float32, Int32})
    @ EaRydCUDA ~/EaRyd/lib/EaRydCUDA/src/opnorm.jl:17
 [20] top-level scope
    @ ~/EaRyd/lib/EaRydCUDA/benchmark/expmv.jl:29
```

**Comment by Roger-luo** [comment]
Ah nvm, I should div the threads

File: lib/cusparse/linalg.jl [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        return
+    end
+
+    m, n = size(A)
+    Tnorm = typeof(float(real(zero(eltype(A)))))
+    Tsum = promote_type(Float64,Tnorm)
+    rowsum = CUDA.CuArray{Tsum}(undef, m)
+    kernel_f = @cuda launch=false kernel(Tnorm, rowsum, A)
+    
+    threads = min(n, CUDA.maxthreads(kernel_f))
```
Comment (maleadt) [inline lib/cusparse/linalg.jl]: Why not use the occupancy API as we normally do?

Comment (Roger-luo) [reply lib/cusparse/linalg.jl]: I won't be able to query threads if I don't have the kernel function object?

Comment (maleadt) [reply lib/cusparse/linalg.jl]: Look for `launch_configuration` in the source code, it also gives you the max threads by using the kernel object.

File: lib/cusparse/linalg.jl [hunk 2]
```diff
@@ -18,7 +18,8 @@ function sum_dim1(A::CuSparseMatrixCSR)
     rowsum = CUDA.CuArray{Tsum}(undef, m)
     kernel_f = @cuda launch=false kernel(Tnorm, rowsum, A)
     
-    threads = min(n, CUDA.maxthreads(kernel_f))
+    config = CUDA.launch_configuration(kernel_f.fun)
```
Comment (maleadt) [inline lib/cusparse/linalg.jl]: `launch_configuration` is exported, so drop the module prefix.

**Comment by maleadt** [comment]
Thanks!

**Pull request merged**